### PR TITLE
Tweaks to IDS

### PR DESCRIPTION
**Pull request by dimven**
Hi @atomczak ,

I'm a colleague of Krzystof and together we used your IDS implementation on a large scale project. We had to make a few tweaks to make it work as desired and I wanted to share them with you. Our files were very large and had a lot of issues, so most of the focus for these changes revolved around reducing the size of the bcf reports to make them easier to manage by the users and to improve the overall performance wherever possible.

- we shifted the outer loop to go through the elements first and then through the specs, so that the elements are iterated only once
- we trimmed the description of the issues
- we fixed a bug where if both the instance and the type have a pset/property with the same name, they get overwritten and the report is wrong.
- we revised the bcf viewport creation, so that the offending elements show up in solibri correctly

Hopefully you'll find some of these changes useful too.

Regards,
Dimitar

### CODE REVIEW

**Comment by Moult** [comment]
Bump and ping @atomczak :)

**Comment by atomczak** [comment]
Hi @dimven! Great! Sorry that I missed your PR, Gmail decided to skip that notification. I will review and test it as soon as I can!

File: src/ifcopenshell-python/ifcopenshell/ids.py [hunk 1]
```diff
@@ -48,52 +48,53 @@ class ids:
...
+        :type creation_date: str, optional
         :param purpose:, defaults to None
         :type purpose: str, optional
         :param milestone:, defaults to None
         :type milestone: str, optional
         """
         self.specifications = []
         self.info = {}
-        if title:
-            self.info["title"] = title
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/ids.py:82]: Why is this deleted?

File: src/ifcopenshell-python/ifcopenshell/ids.py [hunk 2]
```diff
@@ -270,8 +263,7 @@ def asdict(self):
         # if older python collections.OrderedDict()
         spec_dict = {
             "@name": self.name,
-            "@use": self.use,
-            "@ifcVersion": self.ifcVersion,
+            "@necessity": self.necessity,
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/ids.py:266]: I don't see this in either the IDS v4 or v5 XSDs?

File: src/ifcopenshell-python/ifcopenshell/ids.py [hunk 3]
```diff
@@ -1276,52 +1165,42 @@ def emit(self, log_content):
...
         viewpoint = bcf.Viewpoint()
         viewpoint.perspective_camera = bcf.PerspectiveCamera()
-        ifc_elem = log_content.msg["ifc_element"]
-        # ifc_elem = ifc_file.by_guid(log_content.msg["guid"])
+
+        ifc_elem = data["ifc_element"]
         target_position = np.array(ifcopenshell.util.placement.get_local_placement(ifc_elem.ObjectPlacement))
         target_position = target_position[:, 3][0:3]
-        camera_position = target_position + np.array((5, 5, 5))
+        target_position /= 1000 # convert to meters
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/ids.py:1182]: This hardcoded unit conversion doesn't quite look right. Perhaps use something in ifcopenshell.util.unit to help convert?

**Review by Moult — COMMENTED** [review]
Hey @dimven I had a quick look and a lot of the issues seems to be related to going back to IDS v4 instead of IDS v5. Is there a reason for using IDS v4?

**Comment by dimven** [comment]
Hi @Moult 

We started looking into this a few months ago and I guess v4 was the latest back then. This was really not intended as a PR but more of a reference. If you find some of the changes useful, it would be great if you decide the implement them. Otherwise feel free to close this PR. We were trying to apply it on a large-scale project that was generating reports with thousands of issues and we were trying to review the bcfs mainly in Solibri.

One big problem was that the errant elements were not being reported at all and it was because the bcf were being saved with a perspective camera. Each software interprets the perspective cone differently and we were getting broken views with incredibly narrow or distant focal points that led to the exclusion of the elements. Once we switched things to simple orthographic cameras, everything started working as expected.

Another issue was the incredibly long and detailed titles of each issue. Not only was this generating huge reports, but it was also breaking the appearance in Solibri, with the text overflowing over the viewports. It would be great if the final titles were more succinct.

Finally, there were some issues in the processing logic that was misreporting or missing certain property requirements. That we had to work around to get the expected results.

**Comment by aothms** [comment]
> This was really not intended as a PR but more of a reference

That'd really be a shame I think. I haven't followed all of the discussion, but this kind of first hand practical usage experience is really valuable to embed in the code. At the same time we have to make sure the code remains generally applicable. So for example an option to switch between long titles (suitable for text reports?) and short titles (for consumption in UI with limited screen estate?) seems meaningful. In python these kind of options are easily provided using e.g `argparse`. IDS is evolving rapidly so it'd really be a shame if our code bases diverge before IDS v1 is even public. Taking a little bit of time to generalize your insights seems worthwhile in the end and we are here to help with that.

> Finally, there were some issues in the processing logic that was misreporting or missing certain property requirements. That we had to work around to get the expected results.

This seems to be crucial feedback. Can you elaborate?

File: src/ifcopenshell-python/ifcopenshell/ids.py (hunk 2 above)
Comment (atomczak) [reply src/ifcopenshell-python/ifcopenshell/ids.py:266]: 'necessity' was in XSD for some time, but got replaced by 'use'. Now it will probably also be replaced by minOccur in IDS XSD.

**Comment by Moult** [comment]
Closing, as I've reviewed the changes in this PR and I think performance wise the new IfcTester is pretty good and I'm running it on large models in decent time. There are also a number of formatting flexibility improvements with new support for TXT, JSON, HTML, ODS, BCF, and Console outputs. BCF has been updated to the latest BCF v2 library updates too.
